topostats/find_grains.py

```diff
@@ -0,0 +1,220 @@
+"""Find grains in images."""
+# pylint: disable=no-name-in-module
+# pylint: disable=invalid-name
+from pathlib import Path
+import logging
+from typing import Dict, List, Union
+
+import numpy as np
+import pandas as pd
+
+from skimage.filters import gaussian
+from skimage.segmentation import clear_border
+from skimage.morphology import remove_small_objects, label
+from skimage.measure import regionprops
+from skimage.color import label2rgb
+
+from topostats.filters import get_threshold
+from topostats.logs.logs import LOGGER_NAME
+
+LOGGER = logging.getLogger(LOGGER_NAME)
+
+
+def quadratic(x, a, b, c):
+    """Calculate the result of the quadratic equation."""
+    LOGGER.info('Calculating quadratic.')
+    return (a * x**2) + (b * x) + c
+
+
+# Gaussian filtering
+def get_lower_threshold(image: np.array, lower_threshold_multiplier: float = 1.7) -> float:
+    """Calculate the lower threshold for filtering.
+
+    Parameters
+    ----------
+    image: np.array
+        Numpy array representing image.
+    lower_threshold_multiplier: float
+        Factor by which threshold is to be scaled by.
+    Returns
+    -------
+    float
+        Lower threshold
+    """
+    LOGGER.info('Calculating lower threshold for grain identification.')
+    return get_threshold(image) * lower_threshold_multiplier
+
+
+def gaussian_filter(image: np.array, gaussian_size: float = 2, dx: float = 1, mode: str = 'nearest') -> np.array:
+    """Apply Gaussian filter
+
+    Parameters
+    ----------
+    image : np.array
+        Numpy array of image.
+    gaussian_size : float
+        Gaussian blur size in nanometers (nm).
+    dx : float
+        Pixel to nanometer scale.
+    mode : str
+        Mode for filtering (default is 'nearest')
+
+    Returns
+    -------
+    np.array
+        Numpy array of filtered image.
+
+    Examples
+    --------
+    FIXME: Add docs.
+    """
+    LOGGER.info('Applying Gaussian filter (mode : {mode}; Gaussian blur (nm) : {gaussian_size}).')
+    return gaussian(image, sigma=(gaussian_size / dx), mode=mode)
+
+
+def boolean_image(image: np.array, threshold: float) -> np.array:
```

Comment:
Is this much different to `get_mask` in `filters.py`?

Reply:
No its not and I did have a `FIXME` in there to reduce the duplication (which `pylint` complained about), but as I wanted to re-write these as classes I've taken the opportunity to abstract `get_mask()` to `utils.py` and import and use it in both the `Filters` and `FindGrains` classes I've been working on.